sycl/source/device_selector.cpp

```diff
@@ -157,5 +161,152 @@ int host_selector::operator()(const device &dev) const {
   return Score;
 }
 
+namespace ONEAPI {
+namespace detail {
+struct filter {
+  backend Backend;
+  RT::PiDeviceType DeviceType;
+  int DeviceNum;
+  bool hasBackend = false;
+  bool hasDeviceType = false;
+  bool hasDeviceNum = false;
+  int MatchesSeen = 0;
+};
+
+std::vector<std::string> tokenize(std::string filter, std::string delim) {
+  std::vector<std::string> Tokens;
+  size_t Pos = 0;
+  std::string Input = filter;
+  std::string Tok;
+
+  while ((Pos = Input.find(delim)) != std::string::npos) {
+    Tok = Input.substr(0, Pos);
+    Input.erase(0, Pos + delim.length());
+
+    if (!Tok.empty()) {
+      Tokens.push_back(Tok);
+    }
+  }
+
+  // Add remainder
+  if (!Input.empty())
+    Tokens.push_back(Input);
+
+  return Tokens;
+}
+
+filter create_filter(std::string Input) {
+  filter Result;
+  std::string Error = "Invalid filter string! Valid strings conform to "
+                      "BE:DeviceType:DeviceNum, where any are optional";
+
+  std::vector<std::string> Tokens = tokenize(Input, ":");
+  std::regex IntegerExpr("[[:digit:]]+");
+
+  // There should only be up to 3 tokens.
+  // BE:Device Type:Device Num
+  if (Tokens.size() > 3)
+    throw runtime_error(Error, PI_INVALID_VALUE);
+
+  for (const std::string &Token : Tokens) {
+    if (Token == "cpu" && !Result.hasDeviceType) {
+      Result.DeviceType = PI_DEVICE_TYPE_CPU;
+      Result.hasDeviceType = true;
+    } else if (Token == "gpu" && !Result.hasDeviceType) {
+      Result.DeviceType = PI_DEVICE_TYPE_GPU;
+      Result.hasDeviceType = true;
+    } else if (Token == "accelerator" && !Result.hasDeviceType) {
+      Result.DeviceType = PI_DEVICE_TYPE_ACC;
+      Result.hasDeviceType = true;
+    } else if (Token == "opencl" && !Result.hasBackend) {
+      Result.Backend = backend::opencl;
+      Result.hasBackend = true;
+    } else if (Token == "level-zero" && !Result.hasBackend) {
```

Comment:
it is better to use level_zero (underscore instead of minus) as it is common naming for this backend (please update commit message as well)

Reply:
Ok - I just followed our own code: https://github.com/intel/llvm/blob/sycl/sycl/include/CL/sycl/backend_types.hpp

I will change it there too.

Reply:
this is good reason to stick to level-zero

Reply:
Ok

Reply:
Do we prefer "level_zero" (and update the backend_types print) or keep "level-zero"?

Reply:
@smaslov-intel  could you comment what is the best option here.

Reply:
We have used "level_zero" in several places including the backend type. Please use "level_zero".

Reply:
Yes, "level_zero" is the token we've chosen for most uses since we cannot use "-" in type/var names.